164829223-Feature: users highlight article texts and comment on highlights

ogiste opened the pull request:
#### Description
Users can highlight texts from an article and comment on the highlighted text.

#### Type of change
- [ ] Bug fix (nonbreaking change which fixes an issue)
- [x] New feature (nonbreaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

#### How Has This Been Tested?
Please describe the tests that you ran to verify your changes
- [x] End to end
- [x] Integration

#### Checklist:
- [ ] My code follows the style guide for this project
- [x] I have linted my code prior to submission
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

#### How should this be manually tested?
1.  Clone the repository 
```https://github.com/andela/ah-centauri-backend.git```
2. On completing of cloning
```bash 
cd ah-centauri-backend
git checkout feature/164829223-users-highlight-article-texts
```
3. Rename the .env.example file into .env and update the parameters with your own keys.

> For Centauri team members you can use the keys shared through slack

4. Start the development server and send the following through postman

**Get my profile**
```

GET /api/profiles/me/
The current user retrieves their profile information (the profile will now display their highlights and all highlights made on their authored articles)
```

**Get my profile as the current user**
```

GET /api/highlights/
A user retrieves all their highlights
```

**Get all my highlights for an article**
```

GET /api/highlights/<slug:slug>/
A user retrieves all their highlights for an article
```

**Create a highlight for an article**
```

POST /api/highlights/<slug:slug>/
A user creates a highlight or deletes a highlight
```

**POST payload**
```
{
	"highlight_data": {
		"start_index":14,
		"end_index": 23,
		"comment": "New highlight comment"
	}
}
```

**Update the comment of a highlight for an article**
```

PUT /api/highlights/<int:pk>/
A user updates the comment they made to their highlight
```
**PUT payload**
```
{
	"highlight_data": {
		"comment": "New highlight comment"
	}
}
```


#### Pivotal Tracker
[#164829223](https://www.pivotaltracker.com/n/projects/2321088/stories/164829223)

Commit:
164829223-Feature: Users can highlight article text and comment on their highlights-Adds highlights app-Adds Highlights model-Adds Highlights model tests[Starts #164829223]

Commit:
164829223-Feature: Users can highlight article text and comment on their highlights-Adds highlights app-Adds Highlights model-Adds Highlights model tests[Starts #164829223]

ogiste committed:
164829223-Feature: Users can highlight article text and comment on their highlights

-Adds highlights app
-Adds Highlights model
-Adds Highlights model tests
-Add create highlight endpoint
-Add create highlight endpoint tests
-Add delete highlight endpoint
-Add get all my highlights

[Finished #164829223]

ogiste committed:
164829223-Feature: Users can highlight article text and comment on their highlights

-Adds highlights app
-Adds Highlights model
-Adds Highlights model tests
-Add create highlight endpoint
-Add create highlight endpoint tests
-Add delete highlight endpoint
-Add get all my highlights

[Finished #164829223]

ogiste committed:
164829223-Feature: Users can highlight article texts and comment on their highlights

-Adds highlight model and tests
-Adds highlight serializer and tests
-Adds create , delete and get highlights and test
-Adds logic to remove highlights on article body update
-Adds highlights to current user profile

[Finishes #164829223]

koitoror committed:
164829223-Feature: Users can highlight article texts and comment on their highlights

-Adds highlight model and tests
-Adds highlight serializer and tests
-Adds create , delete and get highlights and test
-Adds logic to remove highlights on article body update
-Adds highlights to current user profile

[Delivers #164829223]

ogiste committed:
164829223-Feature: Users can highlight article texts and comment on their highlights

-Adds highlight model and tests
-Adds highlight serializer and tests
-Adds create , delete and get and update highlights views and test
-Adds logic to remove highlights on article body update
-Adds highlights to current user profile

[Finishes #164829223]

ogiste committed:
164829223-Feature: Users can highlight article texts and comment on their highlights

-Adds highlight model and tests
-Adds highlight serializer and tests
-Adds create , delete and get and update highlights views and test
-Adds logic to remove highlights on article body update
-Adds highlights to current user profile

[Finishes #164829223]

Sm0k3s commented on authors/apps/highlights/models.py:
```diff
@@ -0,0 +1,54 @@
+        """
+        username = self.profile.user.username
+        article_title = self.article.title
+        hl_id = self.id
```
use a descriptive variable name, other than that everything looks good.

ogiste replied:
Thank you. Working on this.

bl4ck4ndbr0wn reviewed (approved):
Great work.

oagutu commented on authors/apps/highlights/response_messages.py:
```diff
@@ -0,0 +1,20 @@
+    "HIGHLIGHT_UPDATED": "Highlight has been updated.",
+    "ARTICLE_NOT_FOUND": ("Could not highlight because ",
+                          "the article could not be found."),
+    "HIGHLIGHT_REMOVED": "Highlight removed.",
```
"HIGHLIGHT_UPDATED",  "HIGHLIGHT_ADDED" & "HIGHLIGHT_REMOVED" have been repeated in `HL_VALIDATION_MSGS`

ogiste replied:
Removed duplicates from HL_VALIDATION_MSGS.

oagutu commented on authors/apps/highlights/serializers.py:
```diff
@@ -0,0 +1,93 @@
+        Check that start is less than finishing index and length
+        of article is within index range.
+        """
+        article = self.initial_data['article']
```
This variable is unused

ogiste replied:
Removed unused variable.

oagutu commented on authors/apps/highlights/views.py:
```diff
@@ -0,0 +1,222 @@
+                },
+                status=status.HTTP_404_NOT_FOUND
+            )
+        # Delete the highlight if the start and end index are the same.
```
Why are we deleting as opposed to returning a corresponding message?

ogiste replied:
Given the nature of highlights as a quick and intuitive selection. This would reduce the need for another endpoint just to delete a highlight and create a new one. Front-end logic would allow clicking the highlight icon as the same endpoint to used to create and remove the highlight.

oagutu commented on authors/apps/profiles/models.py:
```diff
@@ -53,6 +56,13 @@ def followers(self):
+    @property
+    def highlights(self):
+        highlights = Highlights.objects.filter(profile=self)
+        if len(highlights) < 1:
```
The if statement here is not necessary as dumping an empty QuerySet into the serialiser will still return an empty list

ogiste replied:
Removed lines of unnecessary code from the highlights property.

ogiste committed:
164829223-Feature: Users can highlight article texts and comment on their highlights

-Adds highlight model and tests
-Adds highlight serializer and tests
-Adds create , delete and get highlights and test
-Adds logic to remove highlights on article body update
-Adds highlights to current user profile

[Delivers #164829223]
